File: src/containers/vector/particles/java/Ensemble.java
/*
 * Algorithms and Complexity                               October 08, 2022
 * IST 4310
 * Prof. M. Diaz-Maldonado
 *
 *
 * Synopsis:
 * Implements the Brute Force and Divide and Conquer Algorithms that solve
 * the Closest Pair problem.
 *
 *
 * Copyright (c) 2022 Misael Diaz-Maldonado
 * This file is released under the GNU General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 *
 * References:
 * [0] A Koenig and B Moo, Accelerated C++ Practical Programming by Example
 * [1] JJ McConnell, Analysis of Algorithms, 2nd edition
 * [2] Sorting Custom Objects Tutorial: (
 *      www.codejava.net/java-core/collections/
 *      sorting-arrays-examples-with-comparable-and-comparator
 * )
 *
 */

import java.util.Random;

public class Ensemble	// Particle Ensemble Class
{
	public static long operations = 0;	// operations counter
	public static double etime = 0;		// elapsed time (nanosecs)

	public static void main (String [] args)
	{
		// defines the system size
		int size = (0x00000100);
		// creates the data set of coordinates without duplicates
		Vector data = genDataSet (size);


		// finds the closest pair with brute force
		double minDistBruteForce = bruteforce (data);


		// initializes operations counter
		operations = 0;
		// finds the closest pair via divide and conquer algorithm
		double minDistRecursive = recursiveClosestPair (data);


		/*

		Reports the test outcome on the console. If the algorithms
		find a closest pair whose distance differ the test fails.
		Note that it is not unlikely for the algorithms to find
		different closest pairs but of equal distance.

		*/


		System.out.printf("test: ");
		if (minDistRecursive > minDistBruteForce)
			System.out.println("FAIL");
		else
			System.out.println("pass");


		// reports system and runtime info on the console
		System.out.printf("size:  %d\n", size);
		System.out.printf("opers: %d\n", operations);
		return;
	}


	public static Vector genDataSet (int size)
	/*

	Synopsis:
	Generates a distinct data set of coordinates by sampling values
	from a uniform pseudo-random number generator PRNG.

	Inputs:
	size		size of the data set (number of particles)

	Output:
	vector		a vector that stores the data set of coordinates

	*/
	{
		// creates a vector for storing the coordinates
		Vector vector = new Vector();
		// creates a pseudo-random number generator PRNG
		Random random = new Random();
		// defines limits for the particle coordinates
		int x_min = -size, x_max = size;
		int y_min = -2, y_max = 2;
		for (int i = 0; i != size; ++i)
		// generates the distinct set of (x, y) coordinates
		{
			int x = x_min + random.nextInt(x_max - x_min);
			int y = y_min + random.nextInt(y_max - y_min);
			Coord c = new Coord (x, y);
			while (vector.search(c) >= 0)
			// generates a new coordinate if already in vector
			{
				x = random.nextInt(size);
				y = random.nextInt(size);
				c = new Coord (x, y);
			}
			// pushes (distinct) coordinate unto back of vector
			vector.push_back(c);
			// sorts to use binary search on next pass
			vector.sort();
		}

		return vector;
	}


	private static double distance (Vector part, int i, int j)
	/*

	Synopsis:
	Returns the squared distance of a pair of particles (i, j) located
	in the same partition.

	Inputs:
	part		partition (or whole data set of coordinates)
	i		index of the ith particle in the left partition
	j		index of the jth particle in the right partition

	Output:
	d		squared distance of the ith and jth particles

	*/
	{
		// gets the coordinate objects of the particles P and Q
		Coord p = part.getData(i), q = part.getData(j);
		// delegates the computation of the squared distance
		return Coord.distance(p, q);
	}


	private static double distance (Vector L, Vector R, int i, int j)
	/*

	Synopsis:
	Returns the squared distance of a pair of particles (i, j) located
	in different partitions.

	Inputs:
	L		left partition coordinates
	R		right partition coordinates
	i		index of the ith particle in the left partition
	j		index of the jth particle in the right partition

	Output:
	d		squared distance of the ith and jth particles

	*/
	{
		// gets the coordinate objects of the particles P and Q
		Coord p = L.getData(i), q = R.getData(j);
		// delegates the computation of the squared distance
		return Coord.distance(p, q);
	}


	private static double distance (Vector part, Coord [] closestPair)
	/*

	Synopsis:
	Applies the Brute Force Algorithm to find the closest pair in a
	partition. Note that the partition could be the whole data set.

	Inputs:
	part		partition (or whole data set of coordinates)
	closestPair	placeholder for storing the closest pair

	Outputs:
	closestPair	the coordinates of the closest pair in partition
	d_min		distance of the closest pair

	*/
	{
		int size = part.size();
		double d_min = Double.POSITIVE_INFINITY;
		for (int i = 0; i != (size - 1); ++i)
		{
			for (int j = (i + 1); j != size; ++j)
			{
				double d = distance (part, i, j);
				if (d < d_min)
				{
					// gets shallow copies of the data
					closestPair[0] = part.getData(i);
					closestPair[1] = part.getData(j);
					d_min = d;
				}
				++operations;
			}
		}

		return d_min;
	}


	private static double distance (Vector L, Vector R, Coord [] pair,
					int [][] arange, double d_min)
	/*

	Synopsis:
	Applies Brute Force Algorithm on the middle partition M.

	Inputs:
	L		left partition
	R		right partition
	pair		current closest pair coordinates
	arange		asymmetric ranges of the left and right partitions
	d_min		current distance of the closest pair

	Outputs:
	pair		most recent coordinates of the closest pair
	d_min		distance of the newest closest pair

	 */
	{
		// gets the asymmetric range of the left partition
		int b1 = arange[0][0], e1 = arange[0][1];
		// gets the asymmetric range of the right partition
		int b2 = arange[1][0], e2 = arange[1][1];


		/*

		NOTE:
		The aranges define which particles in each partition
		could form a closest pair closer than the current one.
		Note that the pairs considered are formed by a particle
		from the left and another particle from right partition,
		for we have already considered pairs within the same
		partition.

		*/


		for (int i = b1; i != e1; ++i)
		{
			for (int j = b2; j != e2; ++j)
			{
				double d = distance(L, R, i, j);
				if (d < d_min)
				{
					// gets shallow copies of the data
					pair[0] = L.getData(i);
					pair[1] = R.getData(j);
					d_min = d;
				}
				++operations;
			}
		}

		return d_min;
	}


	private static double bruteforce (Vector data)
	/*

	Synopsis:
	Applies the Brute Force Algorithm to find the closest pair.

	Input:
	data		coordinates of the particles

	Output:
	d_min		distance of the closest pair

	*/
	{
		// allocates placeholder for storing the closest pair
		Coord [] pair = new Coord[2];
		// applies brute force algorithm on the whole data set
		double d_min = distance (data, pair);

		/*
		Coord [] closestPair = pair;
		// unpacks the (x, y) coordinates of the closest pair
		int x1 = pair[0].getX(), y1 = pair[0].getY();
		int x2 = pair[1].getX(), y2 = pair[1].getY();

		// reports findings on the console
		System.out.printf("\nBrute Force Algorithm:\n");
		System.out.printf("d_min: %f\n", Math.sqrt(d_min));
		System.out.printf("(x1, y1): %2d, %2d\n", x1, y1);
		System.out.printf("(x2, y2): %2d, %2d\n", x2, y2);
		*/
		return Math.sqrt(d_min);
	}


	private static Vector [] divide (Vector Px)
	/*

	Synopsis:
	Divides data set into left and right partitions.

	Input:
	Px		original x-y sorted particle coordinates

	Outputs:
	Lx		left partition
	Rx		right partition

	*/
	{
			// initializes left partition via copy constructor
			Vector Lx = new Vector(Px);
			// creates left and right partitions
			Vector Rx = Lx.bisect();
			// creates placeholder for returning partitions
			Vector [] partitions = {Lx, Rx};
			return partitions;
	}


	private static double select (
		Coord[] closestPair, Coord[] pairLeft, Coord[] pairRight,
		double minDistLeft, double minDistRight
	)
	/*

	Synopsis:
	Selects the closest pair from the left and right partitions.

	Inputs:
	closestPair	placeholder for storing the closest pair
	pairLeft	closest pair in the left partition
	pairRight	closest pair in the right partition
	minDistLeft	distance of the closest pair in the left partition
	minDistRight	distance of the closest pair in the right partition

	Outputs:
	closestPair	the closest of the left and right closest pairs
	minDist		the distance of the new closest pair

	*/
	{
		double minDist;
		if (minDistLeft < minDistRight)
		{
			minDist = minDistLeft;
			// invokes copy constructor to get a shallow copy
			closestPair[0] = new Coord(pairLeft[0]);
			closestPair[1] = new Coord(pairLeft[1]);
		}
		else
		{
			minDist = minDistRight;
			// invokes copy constructor to get a shallow copy
			closestPair[0] = new Coord(pairRight[0]);
			closestPair[1] = new Coord(pairRight[1]);
		}

		/*

		NOTE:
		Generally speaking, it is an error prone practice to get
		a copy of the reference of a local object. This is why we
		opt to copy the object data instead into the placeholder
		that stores the closest pair. Note that the placeholder
		was allocated prior to invoking the recursive method so
		it won't get destroyed until the closest pair is found.
		On the other hand, the placeholders that store the pairs
		in the left and right partitions will get destroyed and
		we could stumble upon a NullPointerException if we get a
		copy of any of their references.

		*/

		return minDist;
	}


	private static double combine (
		Vector Lx, Vector Rx, Coord [] closestPair, double d_min
	)
	/*

	Synopsis:
	Looks for the closest pair at the interface of the left and right
	partitions (dubbed as the middle partition Mx).

	Inputs:
	Lx		left partition
	Rx		right partition
	closestPair	placeholder for the closest pair coordinates
	d_min		distance of the (current) closest pair

	Output:
	d_min		distance of the (possibly new) closest pair

	*/
	{
		/*

		Right Partition Traversal

		Assumes that there are no particles on the right partition
		that could form a pair with any of the particles on the
		left partition which could be closer than the (current)
		closest pair. (Recall that at this point we have already
		considered all possible pairs within the partitions, thus
		we need not to do that again.)

		To keep track of the number of potential candidates we use
		the asymmetric range (arange for short) [b, e), where `b'
		is the beginning limit and `e' is the end limit. The number
		of candidates is computed via the difference (e - b).

		By initializing the range to zero (or empty) we are saying
		that there are no potential closest pair candidates between
		the left and right partitions in accordance with our
		initial assumption.

		The algorithm that follows (for-loop) determines how many
		particles in the right partition could potentially form
		a pair closer than the (current) closest pair.

		Candidate particles on the right partition are selected
		by incremening the end of the arange [b, e). Note that the
		first particle on the right partition is the closest to
		the particles on the left partition and so the arange must
		start with b = 0; also, this is why we only adjust the end
		limit. We update the end limit by traversing the particles
		in the right partition in sequential order. We can afford
		to do that because of the inherent ordering which stems
		from the imposed x-axis sorting.

		If it turns out that there are no candidates, the arange
		remains empty [b, e) = [0, 0), as it should be.

		*/

		// initializes arange [b, e) to empty [0, 0)
		int b2 = 0, e2 = 0;
		for (int i = 0; i != Rx.size(); ++i)
		/*

		Computes the distance between the rightmost particle in the
		left partition and the ith particle on the right partition.
		If the distance of the pair (along the x-axis) is smaller
		than the (current) distance of the closest pair, the ith
		particle is included in the list of (potential) candidates
		by incrementing the end limit of the arange. We stop the
		traversal as soon as there is a pair farther than the
		closest pair, for it is obvious that the next one is going
		to be even farther away owing to the sorting.

		*/
		{
			int last = (Lx.size() - 1);
			int x = Lx.getData(last).getX();
			double dx = (Rx.getData(i).getX() - x);
			if (dx < d_min)
				++e2;
			else
				break;
		}


		/*

		Left Partition Traversal

		As above but the logic is reversed. Again, recall that the
		rightmost particle in the left partition is the closest to
		the right partition and so the end limit is fixed and the
		begin limit of the arange [b, e) gets decremented to add
		particles into the list of closest pair candidates.

		*/

		// initializes arange [b, e) to empty [size, size)
		int b1 = Lx.size(), e1 = Lx.size();
		for (int i = 0; i != Lx.size(); ++i)
		{
			int j = Lx.size() - (i + 1);
			int x = Rx.getData(0).getX();
			double dx = (x - Lx.getData(j).getX());
			if (dx < d_min)
				--b1;
			else
				break;
		}


		/*

		Applies the brute force algorithm to find the closest pair
		in the middle partition. The middle partition consists of
		the possible closest pairs formed by particles in the left
		and right partitions. The number of candidates in the left
		partition is given by (e1 - b1) and those in the right
		partition is given by (e2 - b2). If it turns out that there
		are no candidates the (current) distance of the closest
		is returned.

		*/


		int [][] arange = { {b1, e1}, {b2, e2} };
		return distance(Lx, Rx, closestPair, arange, d_min);
	}


	private static double recurse (Vector Px, Coord [] closestPair)
	/*

	Synopsis:
	Applies the Divide and Conquer Algorithm to find the closest pair.

	Inputs:
	Px		x-y sorted coordinates of the particles
	closestPair	placeholder for storing the closest pair

	Outputs:
	closestPair	coordinates of the closest pair
	minDist		distance of the closest pair

	*/
	{
		if (Px.size() <= 3)
		{
			// uses brute force on the smaller partition
			return distance (Px, closestPair);
		}
		else
		{
			// creates alias
			Coord [] pair = closestPair;
			// allocates (temporary) placeholders
			Coord [] pairLeft  = new Coord[2];
			Coord [] pairRight = new Coord[2];
			// divides domain into left and right partitions
			Vector [] partitions = divide (Px);
			// gets left and right partitions
			Vector Lx = partitions[0], Rx = partitions[1];
			// finds the closest pair in the left partition
			double minDistLeft  = recurse(Lx, pairLeft);
			// finds the closest pair in the right partition
			double minDistRight = recurse(Rx, pairRight);
			// selects the closest from the two partitions
			double minDist = select(pair, pairLeft, pairRight,
						minDistLeft, minDistRight);
			// combines left and right partitions
			return combine(Lx, Rx, pair, minDist);
		}
	}


	public static double recursiveClosestPair (Vector data)
	/*

	Synopsis:
	Applies the Divide and Conquer Algorithm to find the closest pair.

	Input:
	data		coordinates of the particles

	Output:
	d_min		distance of the closest pair

	*/
	{
		Coord [] closestPair = new Coord[2];
		double tstart = System.nanoTime();
		// looks for the closest pair recursively
		double d_min = recurse (data, closestPair);
		double tend = System.nanoTime();

		etime = (tend - tstart);
		/*
		Coord [] pair = closestPair;
		int x1 = pair[0].getX(), y1 = pair[0].getY();
		int x2 = pair[1].getX(), y2 = pair[1].getY();
		System.out.printf("\nDivide and Conquer Algorithm:\n");
		System.out.printf("d_min: %f\n", Math.sqrt(d_min));
		System.out.printf("(x1, y1): %2d, %2d\n", x1, y1);
		System.out.printf("(x2, y2): %2d, %2d\n", x2, y2);
		*/
		return Math.sqrt(d_min);
	}
}


/*
 * COMMENTS:
 * The recursive algorithm could find another closest pair if it turns out
 * that there is more than one closest pair with equal distance. Note that
 * this is not unlikely because the limits imposed on the coordinates of
 * the particles along the y-axis is small. One could increase the limits
 * to try to avoid that but that will be inefficient in terms computations,
 * for there will be partitions where the particles are too far along the
 * y-axis for the partitioning to be meaningful. Note that the algorithm
 * that divides the system into quadrants is better suited in that case.
 *
 */

File: src/algorithms/divide-and-conquer/closest-pair/python/Ensemble.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Closest Pair Lab                                          February 11, 2023

source: Ensemble.py
author: @misael-diaz


Synopsis:
Defines the Point Ensemble Class.


Copyrigth (c) 2023 Misael Diaz-Maldonado
This file is released under the GNU General Public License as published
by the Free Software Foundation, either version 3 of the License, or
at your option) any later version.


References:
[0] JJ McConnell, Analysis of Algorithms, 2nd edition
[1] R Johansson, Numerical Python: Scientific Computing and Data
    Science Applications with NumPy, SciPy, and Matplotlib, 2nd edition

"""


import time                                     # for time_ns() (nanoseconds timer)
import PairUtil                                 # provides min() method for Pair objects
from Pair import Pair                           # provides Pair() constructor
from point import copy                          # provides copy() utility
from Point import Point                         # provides the Point(*) constructor
from Point import contains                      # to find Points by their coordinate data
from numpy.random import randint as random      # to generate integral random numbers
from numpy import array                         # for fast iteration of array slices


from ImplementErrorException import ImplementErrorException


class Ensemble:


    def __init__ (this, size):
        '''
        Synopsis:
        Constructs an ensemble of points of requested size.
        '''

        this._Ensemble__size = copy(int, size)  # sets the number of points
        this._Ensemble__numOperations = 0.0     # zeros the number of operations
        this._Ensemble__elapsedTime = 0.0       # zeros the elapsed-time

        return


    def getSize (this):
        '''
        Synopsis:
        Returns a copy of the ensemble size.
        '''
        return copy(float, this._Ensemble__size)


    def getElapsedTime (this):
        '''
        Synopsis:
        Returns a copy of the elapsed time invested on finding the closest pair.
        '''
        return copy(float, this._Ensemble__elapsedTime)


    def getNumOperations (this):
        '''
        Synopsis:
        Returns a copy of the number of operations invested on finding the closest pair.
        '''
        return copy(float, this._Ensemble__numOperations)


    # methods:


    def bruteForce (this):
        '''
        Synopsis:
        Applies the Brute Force Algorithm to find the closest pair.
        Returns the closest pair.
        '''

        points = this.createDataset1D()
        closestPair = this.timeRecursive1D(points)
        closestPairBruteForce = this.timeBruteForce(points)

        # complains if the closest pairs differ
        if closestPair.compareTo(closestPairBruteForce) != 0:

            print('Brute Force:')
            closestPairBruteForce.print()
            print('Divide And Conquer:')
            closestPair.print()

            err = 'Ensemble.bruteForce(): different closest pairs detected'
            raise ImplementErrorException(err)

        return closestPair


    def recursive1D (this):
        '''
        Synopsis:
        Applies the Divide And Conquer Algorithm to find the closest pair.
        Returns the closest pair.
        '''

        points = this.createDataset1D()
        closestPairBruteForce = this.timeBruteForce(points)
        closestPair = this.timeRecursive1D(points)

        # complains if the closest pairs differ
        if closestPair.compareTo(closestPairBruteForce) != 0:

            print('Brute Force:')
            closestPairBruteForce.print()
            print('Divide And Conquer:')
            closestPair.print()

            err = 'Ensemble.recusive1D(): different closest pairs detected'
            raise ImplementErrorException(err)

        return closestPair


    def timeBruteForce (this, points):
        '''
        Synopsis:
        Times the implementation of the Brute Force Algorithm.
        Sets the elapsed time (nanoseconds) and the number of operations.
        Returns the closest pair.
        '''

        this.isInvalidData(points)

        startTime = time.time_ns()
        closestPair, numOperations = this.distance(points)
        endTime = time.time_ns()

        # sets the elapsed time (nanoseconds):
        elapsedTime = (endTime - startTime)
        this._Ensemble__elapsedTime = elapsedTime

        # sets the number of operations:
        this._Ensemble__numOperations = numOperations

        return closestPair


    def timeRecursive1D (this, points):
        '''
        Synopsis:
        Times the implementation of the 1D Divide And Conquer Algorithm.
        Sets the elapsed time (nanoseconds) and the number of operations.
        Returns the closest pair.
        '''

        this.isInvalidData(points)

        startTime = time.time_ns()
        closestPair, numOperations = this.recurse(0, points.size, points)
        endTime = time.time_ns()

        # sets the elapsed time (nanoseconds):
        elapsedTime = (endTime - startTime)
        this._Ensemble__elapsedTime = elapsedTime

        # sets the number of operations:
        this._Ensemble__numOperations = numOperations

        return closestPair


    def recurse (this, beginPosition, endPosition, Px):
        '''
        Synopsis:
        Implements the 1D Divide And Conquer Algorithm that finds the closest pair.

        Inputs:
        beginPosition       beginning positional index of the virtual partition
        endPosition         ending positional index of the virtual partition

        Px                  first-rank, x-y sorted, numpy, array of Points of size equal
                            to the ensemble size

        Output:
        ret                 returns a tuple containing the closest pair and the number of
                            operations spent on finding the closest pair
        '''

        numel = (endPosition - beginPosition)   # gets the number of elements in partition

        if numel <= 3:
            # applies the Brute Force Algorithm on the small enough partition
            return this.distanceSmallerPartition(beginPosition, endPosition, Px)

        else:

            # divides the virtual partition into smaller left and rigth partitions:

            beginPos = beginPosition
            endPos = beginPosition + (numel // 2)
            # finds the closest pair in the left partition
            dataLeft = this.recurse(beginPos, endPos, Px)
            closestPairLeft, numOperationsLeft = dataLeft

            beginPos = beginPosition + (numel // 2)
            endPos = beginPosition + numel
            # finds the closest pair in the rigth partition
            dataRigth = this.recurse(beginPos, endPos, Px)
            closestPairRigth, numOperationsRigth = dataRigth

            # selects the closest pair from the left and rigth partitions
            closestPair = PairUtil.min(closestPairLeft, closestPairRigth)

            # finds the closest pair at the interface of the left and rigth partitions
            data = this.combine(beginPosition, endPosition, Px, closestPair)
            closestPair, numOperations = data

            # updates the number of operations executed thus far
            numOperations += numOperationsLeft + numOperationsRigth

        return (closestPair, numOperations)


    def combine (this, beginPosition, endPosition, Px, closestPair):
        '''
        Synopsis:
        Finds the closest pair at the interface of the left and rigth partitions.

        Inputs:
        beginPosition       beginning positional index of the virtual partition
        endPosition         ending positional index of the virtual partition

        Px                  first-rank, x-y sorted, numpy, array of Points of size equal
                            to the ensemble size

        Output:
        ret                 returns a tuple containing the closest pair and the number of
                            operations spent on finding the closest pair
        '''


        # gets the squared distance of the current closest pair
        d_min = closestPair.getDistance()


        # defines the asymmetric range that encompasses the left partition
        numel = (endPosition - beginPosition)
        beginPos = beginPosition
        endPos = beginPosition + (numel // 2)

        # assumes there are no closest pair conforming points in the left partition
        beginPosLeft = beginPosition + (numel // 2)
        endPosLeft = beginPosition + (numel // 2)

        # includes points in the left partition comprising closest pair candidates:
        # Note: traverses the left partition from back to front via negative indexing
        for p in Px[ -1 - (Px.size - endPos) : -1 - (Px.size - beginPos) : -1 ]:

            # gets the `x' position of the leftmost point in the rigth partition
            q = Px[endPos]
            x2 = q.getX()

            # gets the `x' position of the (current) rigthmost point in the left partition
            x1 = p.getX()

            # computes the squared x-axis distance of the closest pair candidate
            d = (x2 - x1) * (x2 - x1)

            if d < d_min:
                # includes the point because it can be a closest pair comprising point
                beginPosLeft -= 1
            else:
                # halts the search because the next points are farther away
                break


        # defines the asymmetric range that encompasses the rigth partition
        beginPos = beginPosition + (numel // 2)
        endPos = beginPosition + numel

        # assumes there are no closest pair conforming points in the rigth partition
        beginPosRigth = beginPosition + (numel // 2)
        endPosRigth = beginPosition + (numel // 2)

        # includes points in the rigth partition comprising closest pair candidates:
        for q in Px[ beginPos : endPos ]:

            # gets the `x' position of the rigthmost point in the left partition
            p = Px[beginPos - 1]
            x1 = p.getX()

            # gets the `x' position of the (current) leftmost point in the rigth partition
            x2 = q.getX()

            # computes the squared x-axis distance of the closest pair candidate
            d = (x2 - x1) * (x2 - x1)

            if d < d_min:
                # includes the point because it can be a closest pair comprising point
                endPosRigth += 1
            else:
                # halts the search because the next points are farther away
                break


        # finds the closest pair at the interface of the left and rigth partitions
        data = this.distancePartitionInterface(
            beginPosLeft, endPosLeft, beginPosRigth, endPosRigth, Px, closestPair
        )

        return data


    def distance (this, points):
        '''
        Synopsis:
        Implements the Brute Force Algorithm that finds the closest Pair.
        The theoretical time complexity is O(N**2), where `N' is the size (or length)
        of the array of points.

        Input:
        points              a first-rank, x-y sorted, numpy, array of Points of size equa
                            to the ensemble size

        Output:
        ret                 returns a tuple containing the closest pair and the number of
                            operations spent on finding the closest pair
        '''

        N = points.size
        closestPair = Pair()    # instantiates a Pair of infinitely separated points
        for i, p in enumerate(points[0 : N - 1]):
            for q in points[i + 1 : N]:

                d = p.distance(q)
                pair = Pair(p, q, d)
                closestPair = PairUtil.min(closestPair, pair)

        sz = float(N)
        numOperations = ( sz * (sz - 1) ) / 2

        return (closestPair, numOperations)


    def distanceSmallerPartition (this, beginPosition, endPosition, Px):
        '''
        Synopsis:
        Implements the Brute Force Algorithm that finds the closest Pair in a (smaller)
        partition delimited by the asymmetric range [beginPosition, endPosition).

        Inputs:
        beginPosition       beginning positional index of the smaller partition
        endPosition         ending positional index of the smaller partition

        Px                  first-rank, x-y sorted, numpy, array of Points of size equal
                            to the ensemble size

        Output:
        ret                 returns a tuple containing the closest pair and the number of
                            operations spent on finding the closest pair
        '''

        closestPair = Pair()    # instantiates a Pair of infinitely separated points
        for i, p in enumerate(Px[ beginPosition : endPosition - 1 ]):
            for q in Px[ beginPosition + (i + 1) : endPosition]:

                d = p.distance(q)
                pair = Pair(p, q, d)
                closestPair = PairUtil.min(closestPair, pair)

        numel = float(endPosition - beginPosition)
        numOperations = ( numel * (numel - 1) ) / 2

        return (closestPair, numOperations)


    def distancePartitionInterface (this, beginPosLeft, endPosLeft,
                                    beginPosRigth, endPosRigth, Px,
                                    closestPair):
        '''
        Synopsis:
        Applies the Brute Force Algorithm to find the closest pair at the interface of
        the left and rigth partitions.

        Inputs:
        beginPosLeft        beginning positional index of the smaller left partition
        endPosLeft          ending positional index of the smaller left partition

        beginPosRigth       beginning positional index of the smaller rigth partition
        endPosRigth         ending positional index of the smaller rigth partition

        Px                  first-rank, x-y sorted, numpy, array of Points of size equal
                            to the ensemble size

        Output:
        ret                 returns a tuple containing the closest pair and the number of
                            operations spent on finding the closest pair
        '''

        for p in Px[ beginPosLeft : endPosLeft ]:
            for q in Px[ beginPosRigth : endPosRigth ]:

                d = p.distance(q)
                pair = Pair(p, q, d)
                closestPair = PairUtil.min(closestPair, pair)

        # determines the number of distance computations including those in combine()
        N1 = float(endPosLeft - beginPosLeft)
        N2 = float(endPosRigth - beginPosRigth)
        numOperations = ( (N1 * N2) + (N1 + 1) + (N2 + 1) )

        return (closestPair, numOperations)


    # utilities:


    def createDataset1D (this):
        '''
        Synopsis:
        Creates a dataset of points that has no duplicate closest pairs; that is,
        the points that comprise the second closest pair are farther away than the
        points that comprise the first closest pair.

        Inputs:
        None

        Output:
        dataset             a first-rank, x-y sorted, numpy array of Points of size equal
                            to the ensemble size
        '''

        dataset = this.create()

        while ( this.hasDuplicateClosestPair(dataset) ):
            dataset = this.create()

        return dataset


    def hasDuplicateClosestPair (this, points):
        '''
        Synopsis:
        Returns true if the points that comprise the first and second closest pairs are
        equidistant (duplicate closest pairs); returns false otherwise.

        Input:
        points              a first-rank, x-y sorted, numpy array of Points of size equal
                            to the ensemble size

        Output:
        ret                 true if there is a duplicate closest pair, false otherwise
        '''

        # considers distinct pairs to find the first and second closest pairs:

        N = points.size
        firstClosestPair = Pair()
        secondClosestPair = Pair()
        for i, p in enumerate(points[ 0 : (N - 1) ]):
            for q in points[ (i + 1) : N ]:

                d = p.distance(q)
                pair = Pair(p, q, d)

                if pair.compareTo(firstClosestPair) <= 0:
                    secondClosestPair = firstClosestPair
                    firstClosestPair = pair

        # checks if the closest pairs have equal distances (duplicates):

        d_min = firstClosestPair.getDistance()
        d_2nd = secondClosestPair.getDistance()

        if d_min != d_2nd:
            ret = False
        else:
            ret = True

        return ret


    def isSorted (this, points):
        '''
        Synopsis:
        Returns true if the array of points is x-y sorted, false otherwise.

        Input:
        points              a first-rank, x-y sorted, numpy, array of Points of size equa
                            to the ensemble size

        Output:
        ret                 true if the array is sorted, false otherwise
        '''

        misplacements = 0
        for i in range(points.size - 1):
            # invariant: so far we checked `i' pairs for misplacement (out-of-order)

            p, q = points[i], points[i + 1]

            if p.compareTo(q) > 0:
                misplacements += 1
            else:
                misplacements += 0


        if misplacements != 0:
            return False
        else:
            return True


        return ret


    def isInvalidData (this, points):
        '''
        Synopsis:
        Complains if the array of points is not sorted or if it has duplicates.

        Input:
        points              a first-rank, x-y sorted, numpy, array of Points of size equal
                            to the ensemble size

        Output:
        ret                 true if the array of points is invalid, false otherwise
        '''

        if not this.isSorted(points):
            err = ('Ensemble::isInvalidData(): points must be x-y sorted')
            raise RuntimeError(err)

        if this.hasDuplicates(points):
            err = ('Ensemble::isInvalidData(): points must be distinct')
            raise RuntimeError(err)

        return


    def hasDuplicates (this, dataset):
        '''
        Synopsis:
        Returns true if the array of points has duplicates, false otherwise.

        Input:
        dataset             a first-rank, x-y sorted, numpy, array of Points of size equal
                            to the ensemble size

        Output:
        ret                 true if the array of points has duplicates, false otherwise
        '''

        points = dataset.tolist()
        # sorts (the list just in case it is not sorted) to check for duplicates in pairs
        points.sort(key = lambda p: ( p.getX(), p.getY() ), reverse=False)

        duplicates = 0
        for i in range(len(points) - 1):
            # invariant: so far we checked `i' pairs for equality (duplicate)

            p, q = points[i], points[i + 1]

            if p.compareTo(q) == 0:
                duplicates += 1
            else:
                duplicates += 0


        if duplicates != 0:
            ret = True
        else:
            ret = False


        return ret


    def create (this):
        '''
        Synopsis:
        Creates an ensemble of `N' distinct points, where `N' is the
        ensemble size, by sampling coordinates from a uniformly
        distributed Pseudo-Random Number Generator PRNG.

        Inputs:
        None

        Output:
        points              a first-rank, x-y sorted, numpy array of Points of size equal
                            to the ensemble size
        '''

        sz = copy(int , this._Ensemble__size)
        float_size = copy(float, this._Ensemble__size)
        limit = float_size
        x_min, x_max = [-limit, limit]
        y_min, y_max = [-16, 16]

        # generates an Ensemble of `size' distinct Points:

        points = []
        for i in range(sz):
            x = random(low=x_min, high=x_max, size=None, dtype=int)
            y = random(low=y_min, high=y_max, size=None, dtype=int)
            p = Point(x, y)

            while contains(points, p):
                x = random(low=x_min, high=x_max, size=None, dtype=int)
                y = random(low=y_min, high=y_max, size=None, dtype=int)
                p = Point(x, y)

            points.append(p)

        # sorts the points primarily by their x and secondarily by their y
        # coordinates to support the Divide And Conquer Algorithm

        points.sort(key = lambda p: ( p.getX(), p.getY() ), reverse=False)

        return array(points)
